extensions-core/spark-extensions/src/main/scala/org/apache/druid/spark/v2/writer/DruidDataWriter.scala

```diff
@@ -0,0 +1,304 @@
+          .withQueryGranularity(
+            dataSchema.getGranularitySpec.getQueryGranularity
+          )
+          .withMetrics(dataSchema.getAggregators: _*)
+          .withTimestampSpec(dataSchema.getTimestampSpec)
+          .withMinTimestamp(startMillis)
+          .withRollup(dataSchema.getGranularitySpec.isRollup)
+          .build()
+      )
+      .setMaxRowCount(config.rowsPerPersist)
+      .build()
+  }
+
+  private[v2] def flushIndex(index: IncrementalIndex[_]): IndexableAdapter = {
+    new QueryableIndexIndexableAdapter(
+      closer.register(
+        INDEX_IO.loadIndex(
+          INDEX_MERGER_V9
+            .persist(
+              index,
+              index.getInterval,
+              tmpPersistDir,
+              indexSpec,
+              OnHeapMemorySegmentWriteOutMediumFactory.instance()
+            )
+        )
+      )
+    )
+  }
+
+  private[v2] def getBucketForRow(ts: Long): Long = {
+    dataSchema.getGranularitySpec.getSegmentGranularity.bucketStart(DateTimes.utc(ts)).getMillis
+  }
+
+  override def commit(): WriterCommitMessage = {
+    // Return segment locations on deep storage
+    val specs = bucketToIndexMap.values.map { case (adapters, index) =>
+      if (!index.isEmpty) {
+        adapters += flushIndex(index)
+        index.close()
+      }
+      if (adapters.nonEmpty) {
+        // TODO: Merge adapters up to a total number of rows, and then split into new segments.
+        //  The tricky piece will be determining the partition number for multiple segments (interpolate 0?)
+        val finalStaticIndexer = INDEX_MERGER_V9
+        val file = finalStaticIndexer.merge(
+          adapters.asJava,
+          true,
+          dataSchema.getAggregators,
+          tmpMergeDir,
+          indexSpec,
+          -1 // TODO: Make maxColumnsToMerge configurable
+        )
+        val allDimensions: JList[String] = adapters
+          .map(_.getDimensionNames)
+          .foldLeft(Set[String]())(_ ++ _.asScala)
+          .toList
+          .asJava
+        val interval = adapters.map(_.getDataInterval)
+          .reduce((l, r) => Intervals.utc(
+            Math.min(l.getStartMillis, r.getStartMillis), Math.max(l.getEndMillis, r.getEndMillis)
+          ))
+        val shardSpec = ShardSpecRegistry.createShardSpec(
+          config.shardSpec,
+          partitionMap)
+        val dataSegmentTemplate = new DataSegment(
+          config.dataSource,
+          dataSchema.getGranularitySpec.getSegmentGranularity.bucket(DateTimes.utc(interval.getStartMillis)),
+          config.version,
+          null, // scalastyle:ignore null
+          allDimensions,
+          dataSchema.getAggregators.map(_.getName).toList.asJava,
+          shardSpec,
+          -1,
+          0L
+        )
+        val finalDataSegment = pusher.push(file, dataSegmentTemplate, true)
+        Seq(MAPPER.writeValueAsString(finalDataSegment))
+      } else {
+        Seq.empty
+      }
+    }.toSeq.flatten
+    closer.close()
```

Comment:
I think if this throws an exception we should blow up to avoid polluting the Spark cluster's workspace with orphaned filed, but I'm open to alternative approaches.